### PR TITLE
Cl/testnetdata

### LINKED ISSUE #591: create-testnet-data: --testnet-magic should override the template's value
# Description

When calling `create-testnet-data` and specifying both `--spec-shelley` and `--testnet-magic`, the value of the latter is ignored (which is surprising to users, as reported to me by @andreabedini):

```
→ rm out128 -Rf; cabal run cardano-cli -- conway genesis create-testnet-data --spec-shelley genesis.shelley.json --out-dir out128 --testnet-magic 128 # genesis.shelley.json specifies 42 as the magic
→ cat out128/genesis.json | jq .networkMagic
42
```

whereas if you don't specify a shelley template, `--testnet-magic` is honored:

```
→ rm out128 -Rf; cabal run cardano-cli -- conway genesis create-testnet-data --out-dir out128 --testnet-magic 128
→ cat out128/genesis.json | jq .networkMagic
128
```

The problem lies here: https://github.com/IntersectMBO/cardano-cli/blob/3b70e39299b1279c690ec11cb5aec6c84e13eadc/cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs#L207

```haskell
      Just shelleyPath ->
        newExceptT $ readAndDecodeShelleyGenesis shelleyPath
      Nothing -> do
        -- No template given: a default file is created
        pure $ shelleyGenesisDefaults { sgNetworkMagic = unNetworkMagic (toNetworkMagic networkId) }
```

where the `networkId` (coming from the `--testnet-magic` flag) is only applied in the no `--spec-shelley` case.

# Possible additional work

To be tested if `create-staked` suffer of the same defect.

### PR DESCRIPTION
**Pull request by CarlosLopezDeLara**
# Changelog

```yaml
- description: | 
    This PR improves `create-testnet-data` cmd so that it takes testnet magic either from template file or from the  CLI flag `--testnet-magic`, when used.   
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  - improvement    # QoL changes e.g. refactoring
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

Resolves https://github.com/IntersectMBO/cardano-cli/issues/591

# How to trust this PR

Passing a network magic via the flag overrides the value on the template file
```
carlos@focus:~/GitHub/cardano-cli$ cabal run cardano-cli -- conway genesis create-testnet-data --spec-shelley configuration/cardano/mainnet-shelley-genesis.json --out-dir out128 --testnet-magic 128
Resolving dependencies...
carlos@focus:~/GitHub/cardano-cli$ cat out128/genesis.json | jq .networkMagic
128
```
Using template file without using the --testnet-magic flag takes the value from the template
```
carlos@focus:~/GitHub/cardano-cli$ cabal run cardano-cli -- conway genesis create-testnet-data --spec-shelley configuration/cardano/mainnet-shelley-genesis.json --out-dir outnoflag
Resolving dependencies...
carlos@focus:~/GitHub/cardano-cli$ cat outnoflag/genesis.json | jq .networkMagic
764824073
```
If no template is used the default template is used instead. 

```
cabal run cardano-cli -- conway genesis create-testnet-data --out-dir outdefault
Resolving dependencies...
carlos@focus:~/GitHub/cardano-cli$ cat outdefault/genesis.json | jq .networkMagic42
```

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [x] Self-reviewed the diff

### CODE REVIEW

**Review by carbolymer — APPROVED** [review]
LGTM. Nice!

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs [hunk 1]
```diff
@@ -211,7 +211,7 @@ pGenesisCreateTestNetData envCli =
     <*> pNumUtxoKeys
     <*> pSupply
     <*> pSupplyDelegated
-    <*> pNetworkId envCli
+    <*> (optional $ pNetworkId envCli)
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs]: You should update the parser's help text so the user knows this value will override the genesis file's value.

File: cardano-cli/src/Cardano/CLI/EraBased/Commands/Genesis.hs [hunk 2]
```diff
@@ -88,12 +88,12 @@ data GenesisCreateTestNetDataCmdArgs = GenesisCreateTestNetDataCmdArgs
...
   , numPools :: !Word -- ^ The number of stake pools credentials to create and write to disk.
   , stakeDelegators :: !StakeDelegators -- ^ The number of delegators to pools to create.
-  , numDrepKeys :: !Word -- ^ The number of DRep keys to create. Right now they receive neither delegation nor are registrated. This will come later.
+  , numDrepKeys :: !Word -- ^ The number of DRep keys to create. Right now they receive neither delegation nor are registered. This will come later.
   , numStuffedUtxo :: !Word -- ^ The number of UTxO accounts to make. They are "stuffed" because the credentials are not written to disk.
   , numUtxoKeys :: !Word -- ^ The number of UTxO credentials to create and write to disk.
   , totalSupply :: !(Maybe Lovelace) -- ^ The total number of Lovelace
   , delegatedSupply :: !(Maybe Lovelace) -- ^ The number of Lovelace being delegated
-  , networkId :: !NetworkId -- ^ The network ID to use.
+  , networkId :: !(Maybe NetworkId) -- ^ The network ID to use.
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Commands/Genesis.hs]: Same here e.g "This will override the network id supplied in the genesis file"

**Review by Jimbo4350 — APPROVED** [review]

File: cardano-cli/test/cardano-cli-golden/Test/Golden/CreateTestnetData.hs [hunk 3]
```diff
@@ -62,16 +62,35 @@ tree root = do
...
+-- | This test tests the non-transient case, i.e. it maximizes the files
+-- that can be written to disk.
+golden_create_testnet_data :: ()
+  => Maybe FilePath -- ^ The path to the shelley template use, if any
+  -> Property
+golden_create_testnet_data mShelleyTemplate =
   propertyOnce $ moduleWorkspace "tmp" $ \tempDir -> do
 
+    H.noteIO_ getCurrentDirectory
+
```
Comment (smelc) [inline cardano-cli/test/cardano-cli-golden/Test/Golden/CreateTestnetData.hs]:
```suggestion
```

This was just for debugging and can be removed.

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs (hunk 1 above)
Comment (smelc) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs]: @CarlosLopezDeLara> when you do so, you will need to regenerate the golden files as follows:

```
RECREATE_GOLDEN_FILES=1 cabal test all
```

File: cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs [hunk 4]
```diff
@@ -197,15 +197,23 @@ runGenesisCreateTestNetDataCmd Cmd.GenesisCreateTestNetDataCmdArgs
...
+      Nothing ->
         -- No template given: a default file is created
-        pure $ shelleyGenesisDefaults { sgNetworkMagic = unNetworkMagic (toNetworkMagic networkId) }
+        pure shelleyGenesisDefaults
+  
+  -- Read NetworkId either from file or from the flag. Flag overrides template file.  
+  let actualNetworkId =
+        case networkId of 
+          Just networkFromFlag -> networkFromFlag
+          Nothing -> fromNetworkMagic (NetworkMagic $ sgNetworkMagic shelleyGenesisInit)
```
Comment (smelc) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs]:
```suggestion
```

File: cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs [hunk 5]
```diff
@@ -197,15 +197,23 @@ runGenesisCreateTestNetDataCmd Cmd.GenesisCreateTestNetDataCmdArgs
...
 
-  let -- {0 -> genesis-keys/genesis0/key.vkey, 1 -> genesis-keys/genesis1/key.vkey, ...}
+      shelleyGenesis = shelleyGenesisInit { sgNetworkMagic = unNetworkMagic (toNetworkMagic actualNetworkId) }
+
```
Comment (smelc) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs]:
```suggestion
```

**Comment by CarlosLopezDeLara** [comment]
Addressing @Jimbo4350 comments about updating the parser. 

We were using `pNetworkId :: EnvCli -> Parser NetworkId` which is used by al commands that talk to the node. So instead now we introduce 

```
pGenesisCreateTestNetData :: Parser (GenesisCmds era)
pGenesisCreateTestNetData =
  fmap GenesisCreateTestNetData $ GenesisCreateTestNetDataCmdArgs
    <$> (optional $ pSpecFile "shelley")
...
    <*> optional pNetworkIdForTestnetData
...
  where
    pNetworkIdForTestnetData :: Parser NetworkId 
    pNetworkIdForTestnetData = Testnet . NetworkMagic <$> testnetMagicParser
      where
        testnetMagicParser = Opt.option (bounded "TESTNET_MAGIC")
          (  Opt.long "testnet-magic"
          <> Opt.metavar "NATURAL"
          <> Opt.help "Specify a testnet magic id for the cluster. Overrides the network id supplied in the spec file."
          )
```
This allows for better options, i.e.:  

- Not giving `--mainnet` as an option for create-testnet-data (we want to create testnets, not mainnets),
- Avoid the hanging   `pure <$> maybeToList (envCliNetworkId envCli)`  from pNetworkId which does not make sense here, 
- Showing a better help. In the genesis commands the `networkId` has an entirely different use (setting the magic for the cluster, not for node-to-client)

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs [hunk 6]
```diff
@@ -199,8 +199,8 @@ pGenesisCreateStaked envCli =
     <*> pStuffedUtxoCount
     <*> Opt.optional pRelayJsonFp
 
-pGenesisCreateTestNetData :: EnvCli -> Parser (GenesisCmds era)
-pGenesisCreateTestNetData envCli =
+pGenesisCreateTestNetData :: Parser (GenesisCmds era)
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs]: Now we are no longer able to get the network id from the environment

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]

**Comment by CarlosLopezDeLara** [comment]
With that last changes now we have: 

```
$ cabal run cardano-cli -- conway genesis create-testnet-data --spec-shelley configuration/cardano/mainnet-shelley-genesis.json --out-dir outFromSpecFile
Resolving dependencies...
$ cat outFromSpecFile/genesis.json | jq  .networkMagic
764824073
$ cabal run cardano-cli -- conway genesis create-testnet-data --spec-shelley configuration/cardano/mainnet-shelley-genesis.json --out-dir outFromFlag --testnet-magic 13579
Resolving dependencies...
$ cat outFromFlag/genesis.json | jq  .networkMagic
13579
$ export CARDANO_NODE_NETWORK_ID=24680
$ cabal run cardano-cli -- conway genesis create-testnet-data --spec-shelley configuration/cardano/mainnet-shelley-genesis.json --out-dir outFromEnv 
Resolving dependencies...
$ cat outFromEnv/genesis.json | jq  .networkMagic
24680
$ cabal run cardano-cli -- conway genesis create-testnet-data --spec-shelley configuration/cardano/mainnet-shelley-genesis.json --out-dir outFromFlagOverridesEnv --testnet-magic 9999999 
Resolving dependencies...
$ cat outFromFlagOverridesEnv/genesis.json | jq  .networkMagic
9999999
```
And help menu looks like this: 

```
Usage: cardano-cli conway genesis create-testnet-data [--spec-shelley FILE]
                                                        [--genesis-keys INT]
                                                        [--pools INT]
                                                        [ --stake-delegators INT
                                                        | --transient-stake-delegators INT
                                                        ]
                                                        [--drep-keys INT]
                                                        [--stuffed-utxo INT]
                                                        [--utxo-keys INT]
                                                        [--total-supply LOVELACE]
                                                        [--delegated-supply LOVELACE]
                                                        [--testnet-magic NATURAL]
                                                        [--start-time UTC-TIME]
                                                        --out-dir DIR

  Create data to use for starting a testnet.

Available options:
  ...
  --testnet-magic NATURAL  Specify a testnet magic id for the cluster. This
                           overrides both the network magic from the spec file
                           and CARDANO_NODE_NETWORK_ID environment variable.
...
```

File: cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs [hunk 7]
```diff
@@ -197,15 +197,21 @@ runGenesisCreateTestNetDataCmd Cmd.GenesisCreateTestNetDataCmdArgs
...
-      Nothing -> do
+      Nothing ->
         -- No template given: a default file is created
-        pure $ shelleyGenesisDefaults { sgNetworkMagic = unNetworkMagic (toNetworkMagic networkId) }
-
-  let -- {0 -> genesis-keys/genesis0/key.vkey, 1 -> genesis-keys/genesis1/key.vkey, ...}
+        pure shelleyGenesisDefaults
+  
+  -- Read NetworkId either from file or from the flag. Flag overrides template file.  
+  let actualNetworkId =
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs:209]: In a follow up PR, you can refactor this with [maybe](https://hackage.haskell.org/package/base-4.19.0.0/docs/Prelude.html#v:maybe).

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs [hunk 8]
```diff
@@ -436,3 +435,5 @@ pBulkPoolsPerFile =
     , Opt.help "Each bulk pool to contain this many pool credential sets [default is 0]."
     , Opt.value 0
     ]
+
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs:438]: Random diff

**Review by Jimbo4350 — APPROVED** [review]
LGTM!

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs (hunk 8 above)
Comment (CarlosLopezDeLara) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Genesis.hs:438]: There was an unnecessary empty line on 412. Just deleted that empty line pulling everything bellow one line up.

**Pull request merged**